wp/lib/bap_wp/src/compare.ml

```diff
@@ -113,62 +113,65 @@ let compare_blocks
+          if Bool.is_true (Constr.eval c ctx) then
+            comps, env1, env2
+          else
+            c :: comps, env1, env2)
+  in
+  Constr.mk_clause [] constrs, env1, env2
+
 (* A generic function for generating a predicate that compares two subroutines.
    Takes as input a postcondition generator, and a hypothesis generator.
    this could be made even more generic, say by abstracting over the type of terms
    being compared. *)
 let compare_subs
-    ~postcond:(postcond : comparator)
-    ~hyps:(hyps : comparator)
+    ~postconds:(postconds : comparator list)
+    ~hyps:(hyps : comparator list)
     ~original:(sub1, env1 : Sub.t * Env.t)
     ~modified:(sub2, env2 : Sub.t * Env.t)
-    ~smtlib_post:(smtlib_post : string)
-    ~smtlib_hyp:(smtlib_hyp : string)
   : Constr.t * Env.t * Env.t =
   let env2 = Env.set_freshen env2 true in
   let vars = Var.Set.union
       (Pre.get_vars env1 sub1) (Pre.get_vars env2 sub2) in
   let post, env1, env2 =
-    let post, env1, env2 =
-      postcond ~original:(sub1, env1) ~modified:(sub2, env2) ~rename_set:vars in
-    let smtlib_post = mk_smtlib2_compare env1 env2 smtlib_post in
-    Constr.mk_clause [] [post; smtlib_post], env1, env2
+    fold_comparators postconds ~original:(sub1, env1) ~modified:(sub2, env2)
+      ~rename_set:vars
   in
   info "\nPostcondition:\n%s\n%!" (Constr.to_string post);
   let hyps, env1, env2 =
-    let hyps, env1, env2 =
-      hyps ~original:(sub1, env1) ~modified:(sub2, env2) ~rename_set:vars in
-    let smtlib_hyp = mk_smtlib2_compare env1 env2 smtlib_hyp in
-    let sp_range = Pre.set_sp_range env1 in
-    let init_mem, env1, env2 = init_vars env1 env2 vars in
-    Constr.mk_clause [] (hyps :: smtlib_hyp :: sp_range :: init_mem), env1, env2
+    fold_comparators hyps ~original:(sub1, env1) ~modified:(sub2, env2)
+      ~rename_set:vars
   in
   info "\nHypotheses:\n%s\n%!" (Constr.to_string hyps);
   let pre_mod, env2 = Pre.visit_sub env2 post sub2 in
   let pre_combined, env1 = Pre.visit_sub env1 pre_mod sub1 in
   let goal = Constr.mk_clause [hyps] [pre_combined] in
   goal, env1, env2
 
-let compare_subs_empty
-    ~original:(original : Sub.t * Env.t)
-    ~modified:(modified : Sub.t * Env.t)
-    ~smtlib_post:(smtlib_post : string)
-    ~smtlib_hyp:(smtlib_hyp : string)
-  : Constr.t * Env.t * Env.t =
+let compare_subs_empty () : comparator * comparator =
   let postcond ~original:(_, env1) ~modified:(_,env2) ~rename_set:_ =
     let post = Env.trivial_constr env1 in
     post, env1, env2
   in
   let hyps = postcond in
-  compare_subs ~postcond ~hyps ~original ~modified ~smtlib_post ~smtlib_hyp
+  postcond, hyps
 
-let compare_subs_empty_post
-    ~input:(_ : Var.Set.t)
-    ~original:(original : Sub.t * Env.t)
-    ~modified:(modified : Sub.t * Env.t)
-    ~smtlib_post:(smtlib_post : string)
-    ~smtlib_hyp:(smtlib_hyp : string)
-  : Constr.t * Env.t * Env.t =
+let compare_subs_empty_post () : comparator * comparator =
```

Comment:
Why do all these functions return a pair of `comparator`s?

Reply:
They return comparators so that I could have a list of comparators to pass to `compare_subs` that contain all the hypotheses and postconditions. Would there have been a better way to do this?

Reply:
It's fine. I wonder if it would be worth separating the types of comparators that create hypotheses and ones that create postconditions. But honestly if we don't need to, then it's fine.